Specify all presentation sequences

MDLC01 opened the pull request:
This PR fixes #21, fixes #23, and closes #25, by adding the appropriate variation selectors to symbols that exist in both `sym` and `emoji`. I verified that all the variation sequences are defined in by Unicode.[^1][^2]

[^1]: https://www.unicode.org/reports/tr51/#Emoji_Variation_Sequences
[^2]: https://www.unicode.org/Public/UCD/latest/ucd/emoji/emoji-variation-sequences.txt

I have marked this PR as a draft because the next step is to add variation selectors to all symbols that allow it, whether present in both `sym` and `emoji` or not, to prevent ambiguity.

This made me realize that some emojis are poorly named, but improving this is a task for a separate PR.

Related: https://github.com/typst/typst/pull/6489#discussion_r2211811225.

T0mstone commented:
I'll be the first to bring up automation here:
Whether we want to automate this in the future or not, I think we should definitely have all of these hard-coded in the source code, since such automation would probably take a lot of time that we shouldn't have to pay on each build.
The automation would then only be an optional verification step like in #9. (And maybe we always enable it in CI)

That means any concerns about such automation should not block this PR.

MDLC01 commented:
This PR now contains meta changes.

T0mstone commented:
> For now, I query the list of presentation sequences defined by Unicode from the internet every time the tests are run. Feel free to suggest a better way.

How about caching it in a file `presentation-sequences.txt` and adding `/presentation-sequences.txt` to the `.gitignore`?

Also, I'd like this test to be opt-in to begin with. IMO running `cargo test` shouldn't perform any web requests without the user's consent. A crate feature for this would also have the advantage of making the `reqwest` dependency optional.

T0mstone commented:
> Additionally, I want to clarify that the dependency on `reqwest` is for tests only.

Yeah I got that, but it's still a huge dep tree that not everyone may want to have to download before running the tests.

MDLC01 commented:
I switched to `ureq`. Additionally, the file is now pinned to Unicode 16.0.0 so as to prevent sudden breakage when a new Unicode version releases. The file is now downloaded in `build.rs`. To prevent always having `ureq` as a build dependency, I hid the tests that require it behind a non-default `_test-unicode-conformance` feature and added it to CI.

MDLC01 commented:
According to [The Cargo Book](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies), it is not possible to have a build dependency for tests only, so this feature trick is necessary:

> The same applies to `cfg(debug_assertions)`, `cfg(test)` and `cfg(proc_macro)`. [...] There is currently no way to add dependencies based on these configuration values.

T0mstone commented on build.rs:
```diff
@@ -68,6 +68,18 @@ fn main() {
+
+    #[cfg(feature = "_test-unicode-conformance")]
+    std::fs::write(
+        Path::new(&out).join("emoji-variation-sequences.txt"),
+        ureq::get("https://www.unicode.org/Public/16.0.0/ucd/emoji/emoji-variation-sequences.txt")
```
This still re-dowloads it every time you want to run this test. It might not be much of an issue, but I'd still like it if the file were actually cached in the source tree, especially now that it's pinned to a specific Unicode version, so it doesn't even change usually.
Re-dowloading should only be necessary if the file hasn't been downloaded before or if the cached file is for an older Unicode version (whenever we update that in the future).

T0mstone replied:
I'm still not perfectly happy with this (changes to `src/` shouldn't invalidate the cached file), but I also feel bad for nagging you like this, so I'd be fine with merging as-is and I can just write it myself in a follow-up PR.

MDLC01 replied:
To be clear, I pushed a commit that should make the build script only download the file if it is not already in the Cargo cache. So this shouldn't be an issue anymore.

T0mstone replied:
Alright.
The thing I was worried about was that cargo generates a new `OUT_DIR` sometimes (I have 36 of them for codex alone), but I thought it did this way more often than it apparently does: I just tried changing a bunch of stuff in the source code to trigger a new `OUT_DIR` being generated, but I couldn't get cargo to do it, so I'll have to drop that concern now.

T0mstone commented on src/lib.rs:
```diff
@@ -187,4 +190,152 @@ mod test {
+            .collect::<HashSet<_>>();
+        assert!(
+            are_all_variants_valid(SYM, |c| {
+                c.chars().count() != 1
```
This feels a bit weak; Are you sure there can't be any false positives by just checking that there's more than one codepoint?

MDLC01 replied:
I added a comment to clarify that.

MDLC01 commented on src/lib.rs:
```diff
@@ -285,6 +287,8 @@ mod test {
             .collect::<HashSet<_>>();
         assert!(
             are_all_variants_valid(EMOJI, |c| {
+                // All text presentations are exactly 2 codepoints long as of
```
I changed to "emoji variation sequences" everywhere, as this is how text/emoji presentation sequences seem to be referred to by Unicode.

MDLC01 commented:
Could be `\vs1`–`\vs16`, or alternatively `\evs` and `\tvs` for "emoji variation selector" and "text variation selector" only.[^1]

[^1]: https://www.unicode.org/charts/PDF/UFE00.pdf

MDLC01 commented:
An alternative is to simply paste the corresponding codepoints in the symbol lists

Enivex commented:
> Or `\emoji` and `\text`

I think this would be good

MDLC01 commented:
Since this does not update the public interface, I won't make a PR to typst/typst just to upgrade the commit hash as the upgrade will happen when the new release is made anyway.

laurmaedje commented:
> Since this does not update the public interface, I won't make a PR to typst/typst just to upgrade the commit hash as the upgrade will happen when the new release is made anyway.

I think it couldn't hurt to upgrade it already and add a couple of tests for the variant selector stuff. It should fix some stuff, right?

MDLC01 commented:
I'll do it later then

T0mstone commented:
Just ftr, upgrading the dependency is still blocked on [typst#6489](https://github.com/typst/typst/pull/6489).

laurmaedje commented:
I wasn't sure whether other people were still gonna leave approvals or not, but it didn't seem like it, so I went ahead and just merged. Thanks!
